sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -620,6 +620,37 @@ object PushFoldableIntoBranches extends Rule[LogicalPlan] with PredicateHelper {
   }
 }
 
+/**
+ * Remove duplicated case when branches.
+ */
+object RemoveDuplicatedBranches extends Rule[LogicalPlan] with PredicateHelper {
+
+  /**
+   * Wrapper around a branch that provides semantic equality.
+   */
+  case class EquivalentBranch(br: (Expression, Expression)) {
+    override def equals(o: Any): Boolean = o match {
+      case other: EquivalentBranch if br._1.deterministic && other.br._1.deterministic =>
+        br._1.semanticEquals(other.br._1)
+      case _ => false
+    }
+
+    override def hashCode: Int = br._1.semanticHash()
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case q: LogicalPlan => q transformExpressionsUp {
+      case c @ CaseWhen(branches, _) if branches.length > 1 =>
+        val equivalentBranchSet = branches.map(EquivalentBranch).toSet
+        if (equivalentBranchSet.size < branches.length) {
+          val dedup = equivalentBranchSet.map(_.br).toSeq
```

Comment:
Could this change the order of branches?

Reply:
Good catch. Should be
```scala
 val dedup = branches.map(EquivalentBranch).distinct
 if (dedup.size < branches.length) {
   c.copy(branches = dedup)
 } else {
  c
 }
```